win32ss/user/ntuser/ghost.c

```diff
@@ -128,6 +216,119 @@ HWND FASTCALL UserHungWindowFromGhostWindow(HWND hwndGhost)
     return IntHungWindowFromGhostWindow(pGhostWnd);
 }
 
+BOOL IntHaveToQuitGhosting(void)
+{
+    RTL_ATOM ClassAtom;
+    HWND hwndDesktop;
+    PWND pDesktopWnd;
+    UNICODE_STRING WindowName;
+
+    if (!IntGetAtomFromStringOrAtom(&GhostClass, &ClassAtom))
+        return FALSE;
+
+    hwndDesktop = IntGetCurrentThreadDesktopWindow();
+    pDesktopWnd = ValidateHwndNoErr(hwndDesktop);
+    if (!pDesktopWnd)
+        return FALSE;
+
+    RtlInitUnicodeString(&WindowName, NULL);
+
+    return IntFindWindow(pDesktopWnd, NULL, ClassAtom, &WindowName) != NULL;
+}
+
+VOID NTAPI
+GhostThreadProc(_In_ PVOID StartContext)
+{
+    HWND hwndTarget = (HWND)StartContext;
+    MSG msg;
+
+    if (!IntCreateGhostWindow(hwndTarget))
+    {
+        DPRINT1("Unable to create ghost\n");
+        goto Quit;
+    }
+
+    // thread message loop
+    for (;;)
+    {
+        RtlZeroMemory(&msg, sizeof(MSG));
+
+        if (!co_IntGetPeekMessage(&msg, NULL, 0, 0, PM_REMOVE, TRUE))
+            break;
+
+        switch (msg.message)
+        {
+            case PM_CREATE_GHOST:
+            {
+                hwndTarget = (HWND)msg.wParam;
+                if (!IntCreateGhostWindow(hwndTarget))
+                {
+                    DPRINT1("Unable to create ghost\n");
+                }
+                break;
+            }
+        }
+
+        IntTranslateKbdMessage(&msg, 0);
+        IntDispatchMessage(&msg);
+
+        if (IntHaveToQuitGhosting())
+        {
+            DPRINT1("Have to quit ghost thread\n");
+            break;
+        }
+    }
+
+Quit:
+    RtlFreeHeap(GlobalUserHeap, 0, gGhostInfo);
+
+    KeSetEvent(&gGhostInfo->GhostQuitEvent, IO_NO_INCREMENT, TRUE);
+
+    gGhostInfo = NULL;
+}
+
+BOOL IntAddGhost(HWND hwndTarget)
+{
+    NTSTATUS Status;
+    CLIENT_ID ClientId;
+    HANDLE GhostThreadHandle;
+
+    if (gGhostInfo)
+    {
+        return NtUserPostThreadMessage(gGhostInfo->ThreadId, PM_CREATE_GHOST,
+                                       (WPARAM)hwndTarget, 0);
+    }
+
+    gGhostInfo = RtlAllocateHeap(GlobalUserHeap, HEAP_ZERO_MEMORY, sizeof(GHOST_INFO));
```

Comment:
This should be ExAllocatePoolWithTag with `NonPagedPool`. `KEVENT`s cannot be in paged memory.

Reply:
OK.